Setting up Typescript environment

tidyline opened the pull request:
### What's this "PR" do?
- sorry about messy commit 😉 

#### 1) Upgrade babel version to 7
- using `npx babel-upgrade --write` to upgrade babel version automatically
- moving some packages to `'devDependency'`, which are not needed for production level

#### 2) Setup Typescript
- installing typescript and some `@types` packages, for `index.d.ts` file
  - `@types/react`, `@types/react-dom`, `@types/react-redux`

#### 3) Modify Eslint Config
- ✨newly added typescript eslint config
  -  https://github.com/project-dobby/dobby/pull/11/commits/d5ca4c0b95538b65d95d7c8a7cecba247f0d775a#diff-672e03b5b69a757c8a2ac16b6f926ac2
- seperating eslint config files and 
   overriding 'ts', 'tsx' files for applying 'typescript' lint rules

#### 4) Removing unused Packages
- something like... includes `react-app`

#### 5) Change Folder Structure
<img src="https://user-images.githubusercontent.com/25981942/50557907-edbfc180-0d2c-11e9-8bb5-4d9011b70536.png" width="300" />

- `post` folder is under the `components` folder, 
   and I think files like `reducer.js`, `saga.js`... are not belong to the `components` folder.
- `components` folder only includes `Components`, not "store" or request api things.
- So I changed it like this

<img src="https://user-images.githubusercontent.com/25981942/50557977-86564180-0d2d-11e9-818a-2b65f04fefea.png" width="300" />
<img src="https://user-images.githubusercontent.com/25981942/50557983-8a825f00-0d2d-11e9-92a7-d81676e77c09.png" width="300" />




### Points to review.
#### Eslint
- I've tried to set eslint config in the same('.eslintrc') file, 
  but I don't know how to apply rules to only `.ts`, `.tsx` file not sparating files.
- I've changed `.eslintrc.js` to `.eslintrc`.
  It there any reason using `.js` file?

#### My Apology
- I couldn't reviewed all options of eslint config and typescript config 😢 
  So I temporarily added some options not knowing correctly.
  I will fix and refine those options on the next PR. 😭 

### Any background context you want to provide?
- N/A

### Screenshots (if appropriate)
- N/A

tidyline committed:
:arrow_up: npx babel-upgrade --write

tidyline committed:
:heavy_minus_sign: yarn remove babel-preset-react-app

babel-preset-react-app's babel version is not fully matched with upgraded babel version, 7.

tidyline committed:
:heavy_plus_sign: yarn add --dev @babel/preset-react

tidyline committed:
:heavy_plus_sign: yarn add --dev typescript

tidyline committed:
:heavy_plus_sign: yarn add --dev @types/react

tidyline committed:
:heavy_plus_sign: yarn add --dev @babel/react-dom

tidyline committed:
:heavy_plus_sign: yarn add --dev @types/react-redux

tidyline committed:
:sparkles: apply typescript and modify lint config

yeonkko commented on .eslintrc:
```diff
@@ -0,0 +1,27 @@
+{
```
https://eslint.org/docs/user-guide/configuring#configuration-file-formats
`.eslintrc` File format was deprecated

tidyline replied:
Oh I didn't know about this.. 😄 I fixed it on commit https://github.com/project-dobby/dobby/pull/11/commits/63174bc4b9cc8b71a7b954c6796093135272ac5b

tidyline committed:
:ok_hand: restore .eslintrc to .eslintrc.js

cnabro commented on config/jest/fileTransform.js:
```diff
@@ -6,5 +6,5 @@ const path = require("path");
 module.exports = {
   process(src, filename) {
     return `module.exports = ${JSON.stringify(path.basename(filename))};`;
-  }
```
Some of files is changed to using comma seperator on last file. Is there any difference?
Or just rules of lint?

tidyline replied:
I changed `package.json` file's `lint` command like [this](https://github.com/project-dobby/dobby/pull/11/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R23).
So some of files were newly included to eslint target.

yeonkko commented:
how about use [tslint](https://github.com/palantir/tslint) instead eslint :)

tidyline committed:
:heavy_plus_sign: yarn add --dev tslint

tidyline committed:
:fire: remove typescript eslint rules

tidyline committed:
:heavy_plus_sign: yarn add --dev tslint-react

tidyline committed:
:sparkles: add tslint command

tidyline committed:
:wrench: yarn tslint

tidyline commented:
@yeonkko 
I added  [`tslint`](https://www.npmjs.com/package/tslint) and [`tslint-react`](https://github.com/palantir/tslint-react) preset, but I didn't remove eslint for `.js` file.
If we modify some of `.js` files, I think `eslint` also needed :)

Some of typescript rules are not match with our project, so I modified some rules.
- [no-submodule-imports](https://palantir.github.io/tslint/rules/no-submodule-imports/)
- [no-implicit-dependencies](https://palantir.github.io/tslint/rules/no-implicit-dependencies/)
  - set `true` not to trigger errors when import like this, `'src/a/b'`. 
  - set `src` as dependencies
- [trailing-comma](https://palantir.github.io/tslint/rules/trailing-comma/)
  - allow trailing comma
- [object-literal-sort-keys](https://palantir.github.io/tslint/rules/object-literal-sort-keys/)
  - if it is `true`, object literal must be like this, `var a = {a, b, c}`, not `var a = {b, a, c}`;
